rali/rali/source/tf_record_reader.cpp

```diff
@@ -0,0 +1,329 @@
+    if(!file_contents)
+        THROW("TFRecordReader: Failed to open file "+fname);
+    file_contents.seekg(0, std::ifstream::end);
+    file_size = file_contents.tellg();
+    // std::cerr<<"\n length of the file:: "<<length<<std::endl;
+    file_contents.seekg(0, std::ifstream::beg);
+    auto ret = read_image_names(file_contents, file_size);
+    if(ret != Reader::Status::OK )
+        THROW("TFRecordReader: Error in reading TF records");
+    _last_rec = false;
+    if (_file_names.size() != _file_size.size())
+        std::cerr << "\n Size of vectors are not same";
+    file_contents.close();
+
+    return Reader::Status::OK;
+}
+
+size_t TFRecordReader::get_file_shard_id()
+{
+    if (_batch_count == 0 || _shard_count == 0)
+        THROW("Shard (Batch) size cannot be set to 0")
+    return (_file_id / (_batch_count)) % _shard_count;
+}
+
+Reader::Status TFRecordReader::read_image_names(std::ifstream &file_contents, uint file_size)
+{
+    auto ret = Reader::Status::OK;
+    uint length;
+    size_t uint64_size, uint32_size;
+    uint64_t data_length;
+    uint32_t length_crc, data_crc;
+    uint64_size = sizeof(uint64_t);
+    uint32_size = sizeof(uint32_t);
+    while (!_last_rec)
+    {
+        length = file_contents.tellg();
+        char *header_length = new char[uint64_size];
+        char *header_crc = new char[uint32_size];
+        char *footer_crc = new char[uint32_size];
+        file_contents.read(header_length, uint64_size);
+        if(!file_contents)
+            THROW("TFRecordReader: Error in reading TF records")
+        file_contents.read(header_crc, uint32_size);
+        if(!file_contents)
+            THROW("TFRecordReader: Error in reading TF records")
+        memcpy(&data_length, header_length, sizeof(data_length));
+        memcpy(&length_crc, header_crc, sizeof(length_crc));
+        if (uint(length + data_length + 16) == file_size)
+        {
+            _last_rec = true;
+        }
+        char *data = new char[data_length];
+        file_contents.read(data, data_length);
+        if(!file_contents)
+            THROW("TFRecordReader: Error in reading TF records")  
+        _single_example.ParseFromArray(data, data_length);
+        _features = _single_example.features();
+        auto feature = _features.feature();
+        _single_feature = feature.at("image/filename");
+        std::string fname = _single_feature.bytes_list().value()[0];
+        _image_record_starting.insert(std::pair<std::string, uint>(fname, length));
+        _in_batch_read_count++;
+        _in_batch_read_count = (_in_batch_read_count % _batch_count == 0) ? 0 : _in_batch_read_count;
+        std::string file_path = _folder_path;
+        file_path.append("/");
+        file_path.append(fname);
+        _last_file_name = file_path;
+        if (get_file_shard_id() != _shard_id)
+        {
+            incremenet_file_id();
+            file_contents.read(footer_crc, sizeof(data_crc));
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5763058)

Reply:
This is fstream read. We are evaluating the read status in the next line.
if(!file_contents)
THROW("TFRecordReader: Error in reading TF records")
This can be ignored.